src/python/deepgnn/graph_engine/snark/decoders.py

```
@@ -43,6 +43,219 @@ def decode(self, line: str) -> Iterator[Tuple[int, int, int, float, list]]:
+    """
+
+    def __init__(
+        self,
+        default_node_type: Optional[int] = None,
+        default_edge_type: Optional[int] = None,
+        default_node_weight: Optional[float] = None,
+        default_edge_weight: Optional[float] = None,
+        default_node_feature_types: Optional[List[Optional[str]]] = None,
+        default_edge_feature_types: Optional[List[Optional[str]]] = None,
+        default_node_feature_lens: Optional[List[Optional[List[int]]]] = None,
+        default_edge_feature_lens: Optional[List[Optional[List[int]]]] = None,
+        delimiter=",",
+        length_delimiter="/",
+    ):
+        """Initialize the Decoder."""
+        super().__init__()
+        self.node_type = default_node_type
+        self.node_weight = default_node_weight
+        self.node_feature_types = (
+            default_node_feature_types if default_node_feature_types is not None else []
+        )
+        self.node_feature_lens = (
+            default_node_feature_lens if default_node_feature_lens is not None else []
+        )
+        self.n_node_feature = len(self.node_feature_types)
+
+        self.edge_type = default_edge_type
+        self.edge_weight = default_edge_weight
+        self.edge_feature_types = (
+            default_edge_feature_types if default_edge_feature_types is not None else []
+        )
+        self.edge_feature_lens = (
+            default_edge_feature_lens if default_edge_feature_lens is not None else []
+        )
+        self.n_edge_feature = len(self.edge_feature_types)
+
+        self.delimiter = delimiter
+        self.length_delimiter = length_delimiter
+        self.escape = r"\\"[0]
+
+    def _get_feature(
+        self, key: str, length: List[int], data: Iterator
+    ) -> Union[np.ndarray, Tuple[np.ndarray, np.ndarray], None]:
+        def _str_get(item, idx):
+            if not len(item):
+                return ""
+            return item[idx]
+
+        try:
```

I still don't understand what this code does:
1. Why do you want to convert to int and immediately throw an exception? and later use key as a `dtype` to create arrays. Both strings `binary` and `some other random string` will raise `ValueError` if you'll try to convert it to int.
2. Where do you check `Expected feature vector key is str.` message in tests? If the error is silently swallowed inside the code, why do we need it?